flashbax/vault/vault.py

```diff
@@ -0,0 +1,448 @@
+                target=experience_structure,
+            )
+
+        # Each leaf of the fbx_state.experience maps to a data store, so we tree map over the
+        #  tree structure to create each of the data stores.
+        self._all_ds = jax.tree_util.tree_map_with_path(
+            lambda path, x: self._init_leaf(
+                name=_path_to_ds_name(path),
+                leaf=x,
+                create_ds=not base_path_exists,
+            ),
+            self._tree_structure,
+            # Tree structure uses a list [shape, dtype] as a leaf
+            is_leaf=lambda x: isinstance(x, list),
+        )
+
+        # We keep track of the last fbx buffer idx received
+        self._last_received_fbx_index = 0
+
+        # We store and load the vault index from a separate datastore
+        self._vault_index_ds = ts.open(
+            self._get_base_spec("vault_index"),
+            dtype=jnp.int32,
+            shape=(1,),
+            create=not base_path_exists,
+        ).result()
+        # Just read synchronously as it's one number
+        self.vault_index = int(self._vault_index_ds.read().result()[0])
+
+    def _get_base_spec(self, name: str) -> dict:
+        """Simple common specs for all datastores.
+
+        Args:
+            name (str): name of the datastore
+
+        Returns:
+            dict: config for the datastore
+        """
+        return {
+            "driver": "zarr",
+            "kvstore": {
+                "driver": "ocdbt",
+                "base": f"{DRIVER}{self._base_path}",
+                "path": name,
+            },
+        }
+
+    def _init_leaf(
+        self, name: str, leaf: list, create_ds: bool = False
+    ) -> ts.TensorStore:
+        """Initialise a datastore for a leaf of the experience tree.
+
+        Args:
+            name (str): datastore name
+            leaf (list): leaf of the form ["shape", "dtype"]
+            create_ds (bool, optional): whether to create the datastore. Defaults to False.
+
+        Returns:
+            ts.TensorStore: the datastore object
+        """
+        spec = self._get_base_spec(name)
+        # Convert shape and dtype from str to tuple and dtype
+        leaf_shape = make_tuple(leaf[0])  # Must convert to a real tuple
+        leaf_dtype = leaf[1]  # Can leave this as a str
+
+        leaf_ds = ts.open(
+            spec,
+            # Only specify dtype and shape if we are creating a vault
+            # (i.e. don't impose dtype and shape if we are _loading_ a vault)
+            dtype=leaf_dtype if create_ds else None,
+            shape=(
+                leaf_shape[0],  # Batch dim
+                TIME_AXIS_MAX_LENGTH,  # Time dim, which we extend
+                *leaf_shape[2:],  # Experience dim(s)
+            )
+            if create_ds
+            else None,
+            # Only create datastore if we are creating the vault:
+            create=create_ds,
+        ).result()  # Do this synchronously
+
+        return leaf_ds
```

Comment:
Slightly cleaner to define shape and dtype outside
```suggestion
        dtype, shape = None, None
        
        # Only specify dtype and shape if we are creating a vault
        # (i.e. don't impose dtype and shape if we are _loading_ a vault)
        if create_ds:
          dtype = leaf[1] # Can leave this as a str
          shape = (
                leaf_shape[0],  # Batch dim
                TIME_AXIS_MAX_LENGTH,  # Time dim, which we extend
                *leaf_shape[2:],  # Experience dim(s)
            )
        
        leaf_ds = ts.open(
            spec,
            dtype=dtype,
            shape=shape,
            # Only create datastore if we are creating the vault:
            create=create_ds,
        ).result()  # Do this synchronously
```

Reply:
yes! done 💯